hack/cmd/teststale/teststale.go

```diff
@@ -0,0 +1,203 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+var (
+	binary  = flag.String("binary", "", "absolute filesystem path to the test binary")
+	pkgPath = flag.String("package", "", "test package import path in the format used in the import statements without the $GOPATH prefix")
+)
+
+type pkg struct {
+	dir          string
+	target       string
+	stale        bool
+	testGoFiles  []string
+	testImports  []string
+	xTestGoFiles []string
+	xTestImports []string
+}
+
+func newCmd(format string, pkgPaths []string) *exec.Cmd {
+	args := []string{
+		"list",
+		"-f",
+		format,
+	}
+	args = append(args, pkgPaths...)
+	cmd := exec.Command("go", args...)
+	cmd.Env = os.Environ()
+	return cmd
+}
+
+func newPkg(path string) (*pkg, error) {
+	format := "Dir: {{println .Dir}}Target: {{println .Target}}Stale: {{println .Stale}}TestGoFiles: {{println .TestGoFiles}}TestImports: {{println .TestImports}}XTestGoFiles: {{println .XTestGoFiles}}XTestImports: {{println .XTestImports}}"
+	cmd := newCmd(format, []string{path})
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("could not pipe STDOUT: %v", err)
+	}
+
+	// Start executing the command
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("command did not start: %v", err)
+	}
+
+	// Parse the command output
+	scanner := bufio.NewScanner(stdout)
+	scanner.Split(bufio.ScanLines)
+
+	// To be conservative, default to package to be stale
+	p := &pkg{
+		stale: true,
+	}
+
+	// TODO: avoid this stupid code repetition by iterating through struct fields.
+	scanner.Scan()
+	p.dir = strings.TrimPrefix(scanner.Text(), "Dir: ")
+	scanner.Scan()
+	p.target = strings.TrimPrefix(scanner.Text(), "Target: ")
+	scanner.Scan()
+	if strings.TrimPrefix(scanner.Text(), "Stale: ") == "false" {
+		p.stale = false
+	}
+	p.testGoFiles = scanLineList(scanner, "TestGoFiles: ")
+	p.testImports = scanLineList(scanner, "TestImports: ")
+	p.xTestGoFiles = scanLineList(scanner, "XTestGoFiles: ")
+	p.xTestImports = scanLineList(scanner, "XTestImports: ")
+
+	if err := cmd.Wait(); err != nil {
+		return nil, fmt.Errorf("command did not complete: %v", err)
+	}
+	return p, nil
+}
+
+func (p *pkg) isStale(buildTime time.Time) bool {
+	// If the package itself is stale, then we have to rebuild the whole thing anyway.
+	if p.stale {
+		return true
+	}
+
+	// Test for file staleness
+	for _, f := range p.testGoFiles {
+		if isStale(buildTime, filepath.Join(p.dir, f)) {
+			log.Printf("test Go file %s is stale", f)
+			return true
+		}
+	}
+	for _, f := range p.xTestGoFiles {
+		if isStale(buildTime, filepath.Join(p.dir, f)) {
+			log.Printf("external test Go file %s is stale", f)
+			return true
+		}
+	}
+
+	format := "{{.Stale}}"
```

Comment:
`go list -f '{{.Stale}}' pkg` does a staleness check on the package by checking the package's non-test `.go` files and all the dependencies recursively. It doesn't check the dependencies imported by the `_test.go` files. I run `go list` the first time to get those dependencies in `p.testImports` and `p.xTestImports`. And then I run `go list` again on those packages.